src/dragonball/src/vm/x86_64.rs

```diff
@@ -0,0 +1,279 @@
+            )
+            .map_err(Error::BootSystem)?;
+        }
+    }
+
+    let zero_page_addr = GuestAddress(layout::ZERO_PAGE_START);
+    guest_mem
+        .checked_offset(zero_page_addr, mem::size_of::<bootparam::boot_params>())
+        .ok_or(Error::ZeroPagePastRamEnd)?;
+    guest_mem
+        .write_obj(params, zero_page_addr)
+        .map_err(|_| Error::ZeroPageSetup)?;
+
+    Ok(())
+}
+
+impl Vm {
+    /// Get the status of in-kernel PIT.
+    pub fn get_pit_state(&self) -> Result<kvm_pit_state2> {
+        self.vm_fd
+            .get_pit2()
+            .map_err(|e| Error::Vm(VmError::Irq(e)))
+    }
+
+    /// Set the status of in-kernel PIT.
+    pub fn set_pit_state(&self, pit_state: &kvm_pit_state2) -> Result<()> {
+        self.vm_fd
+            .set_pit2(pit_state)
+            .map_err(|e| Error::Vm(VmError::Irq(e)))
+    }
+
+    /// Get the status of in-kernel ioapic.
+    pub fn get_irqchip_state(&self, chip_id: u32) -> Result<kvm_irqchip> {
+        let mut irqchip: kvm_irqchip = kvm_irqchip {
+            chip_id,
+            ..kvm_irqchip::default()
+        };
+        self.vm_fd
+            .get_irqchip(&mut irqchip)
+            .map(|_| irqchip)
+            .map_err(|e| Error::Vm(VmError::Irq(e)))
+    }
+
+    /// Set the status of in-kernel ioapic.
+    pub fn set_irqchip_state(&self, irqchip: &kvm_irqchip) -> Result<()> {
+        self.vm_fd
+            .set_irqchip(irqchip)
+            .map_err(|e| Error::Vm(VmError::Irq(e)))
+    }
+}
+
+impl Vm {
+    /// Initialize the virtual machine instance.
+    ///
+    /// It initialize the virtual machine instance by:
+    /// 1) initialize virtual machine global state and configuration.
+    /// 2) create system devices, such as interrupt controller, PIT etc.
+    /// 3) create and start IO devices, such as serial, console, block, net, vsock etc.
+    /// 4) create and initialize vCPUs.
+    /// 5) configure CPU power management features.
+    /// 6) load guest kernel image.
+    pub fn init_microvm(
+        &mut self,
+        epoll_mgr: EpollManager,
+        vm_as: GuestAddressSpaceImpl,
+        request_ts: TimestampUs,
+    ) -> std::result::Result<(), StartMicroVmError> {
+        info!(self.logger, "VM: start initializing microvm ...");
+
+        self.init_tss()?;
+        // For x86_64 we need to create the interrupt controller before calling `KVM_CREATE_VCPUS`
+        // while on aarch64 we need to do it the other way around.
+        self.setup_interrupt_controller()?;
+        self.create_pit()?;
+        self.init_devices(epoll_mgr)?;
+
+        let reset_event_fd = self.device_manager.get_reset_eventfd().unwrap();
+        self.vcpu_manager()
+            .map_err(StartMicroVmError::Vcpu)?
+            .set_reset_event_fd(reset_event_fd)
+            .map_err(StartMicroVmError::Vcpu)?;
+
+        if self.vm_config.cpu_pm == "on" {
+            info!(self.logger, "VM: enable CPU disable_idle_exits capability");
+        }
```

Comment:
- What values can `cpu_pm` take? It looks rather odd to be checking a string value `"on"`.
- This doesn't seem to do anything apart from make a log call :) Will code be added in a follow-up PR?

Reply:
#4590